src/beamlime/communication/pipes.py

```diff
@@ -0,0 +1,238 @@
+
+
+class SyncBuffer(BufferBase[BufferData]):
+    """Synchronous reading interfaces of a buffer."""
+
+    def readall(self) -> Generator[BufferData, None, None]:
+        """
+        Returns an iterator that yields the first piece of data
+        until the buffer is empty.
+        """
+        while self._data:
+            yield self._data.pop(0)
+
+    def read(self) -> Union[BufferData, Type[Empty]]:
+        """
+        Pop the first piece of data if there is any data available,
+        else return ``queue.Empty``.
+        """
+        try:
+            return self._data.pop(0)
+        except IndexError:
+            return Empty
+
+
+class AsyncBuffer(BufferBase[BufferData]):
+    """Asynchronous reading interfaces of a buffer."""
+
+    async def readall(self) -> AsyncGenerator[BufferData, None]:
+        """
+        Returns an iterator that yields the first piece of data
+        until the buffer is empty.
+        """
+        while self._data:
+            yield self._data.pop(0)
+
+    async def read(self) -> Union[BufferData, Type[Empty]]:
+        """
+        Pop the first piece of data if there is any data available,
+        else return ``queue.Empty``.
+        """
+        try:
+            return self._data.pop(0)
+        except IndexError:
+            return Empty
+
+
+BufferType = Union[Type[SyncBuffer[BufferData]], Type[AsyncBuffer[BufferData]]]
+Buffer = Union[SyncBuffer[BufferData], AsyncBuffer[BufferData]]
+
+
+class Pipe(BufferBase[BufferData]):
+    """
+    Pipe object can create a buffer to read a data from the pipe,
+    or write a chunk or a piece of data into the pipe.
+
+    If there is an exception raised in the reading data context,
+    it will restore all data from the chunk into the pipe.
+
+    If the context reaches the end without any exception raised,
+    consumed data will be discarded from the buffer
+    and the rest will be restored into the pipe.
+    """
+
+    def __init__(
+        self,
+        *_initial_data: BufferData,
+        max_size: int = MAX_BUFFER_SIZE,
+        chunk_size: int = DEFAULT_CHUNK_SIZE,
+    ) -> None:
+        super().__init__(*_initial_data, max_size=max_size)
+        self.chunk_size: int = chunk_size
+
+    @property
+    def chunk_size(self) -> int:
+        return self._chunk_size
+
+    @chunk_size.setter
+    def chunk_size(self, _chunk_size: int) -> None:
+        if _chunk_size > MAX_CHUNK_SIZE:
+            raise ValueError("Maximum chunk size is limited to ", MAX_CHUNK_SIZE)
+        self._chunk_size = _chunk_size
+
+    @contextmanager
+    def _open_buffer(
+        self, buffer_type: BufferType[BufferData]
+    ) -> Iterator[Buffer[BufferData]]:
+        """
+        Create a buffer of the ``buffer_type``.
+        """
+        try:
+            from copy import deepcopy
+
+            chunk, self._data = (
+                self._data[: self.chunk_size],
+                self._data[self.chunk_size :],
+            )
```

Comment:
So if I understand correctly, `Buffer` and its subclasses support some way of chunking data for some purpose. It feels like an odd and indirect interface. Can you elaborate a bit on where this is needed, and why you chose this solution?